WebContent Process capture thread can become de-prioritized; leading to underruns

jernoble opened the pull request:
#### ada13b119831c938d4a5fbbd5c502c3d3ae6c667
<pre>
WebContent Process capture thread can become de-prioritized; leading to underruns
<a href="https://bugs.webkit.org/show_bug.cgi?id=293295">https://bugs.webkit.org/show_bug.cgi?id=293295</a>
<a href="https://rdar.apple.com/problem/151699897">rdar://problem/151699897</a>

Reviewed by Youenn Fablet.

The Audio IO thread in the GPU process is running at a realtime scheduling priority,
reading samples from a shared ring buffer for Web Audio, and storing samples into a
shared ring buffer for microphone capture. While the matching thread storing samples
for Web Audio in the Web Content process (owned by RemoteAudioDestinationManager) is
also running at a realtime priority, the matching thread reading microphone samples
from the shared ring buffer is running at a normal thread priority. This allows the
kernel to deprioritize and preempt that thread, which can lead to dropped microphone
samples when used as an input to Web Audio.

For now, the limit to the number of concurrent realtime audio threads is 3. Given
that its relatively unlikely to have three capture sessions running simultaneously
this is likely good enough, and in the future, we can and should coalesce all of
of these threads into a single audio thread.

* Source/WTF/wtf/Threading.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
(WebKit::RemoteCaptureSampleManager::RemoteAudio::stopThread):
(WebKit::RemoteCaptureSampleManager::RemoteAudio::startThread):
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.h:

Canonical link: <a href="https://commits.webkit.org/295769@main">https://commits.webkit.org/295769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db00d048774d3cbbfc87ef4d3799441e5d684cc2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16107 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111160 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56560 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108003 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34216 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80495 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20778 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95630 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60815 "Found 142 new API test failures: /WPE/TestUIClient:/webkit/WebKitWebView/query-permission-requests, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/deviceidhashsalt, /WPE/TestLoaderClient:/webkit/WebKitWebView/load-twice-and-reload, /WPE/TestCookieManager:/webkit/WebKitCookieManager/ephemeral, /WPE/TestWebKitWebView:/webkit/WebKitWebView/is-playing-audio, /WPE/TestConsoleMessage:/webkit/WebKitConsoleMessage/network-error, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration, /WPE/TestCookieManager:/webkit/WebKitCookieManager/delete-cookies, /WPE/TestWebKitWebView:/webkit/WebKitWebView/terminate-unresponsive-web-process, /WPE/TestSSL:/webkit/WebKitWebView/insecure-content ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13727 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55998 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98603 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90159 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13763 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114012 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104581 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33102 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24424 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89576 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33466 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91858 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89256 "Found 101 new API test failures: /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /TestWebKit:WebKit.AboutBlankLoad, /TestWebKit:WebKit.PreventEmptyUserAgent, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/pointer-lock-permission-request, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit.Find, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebProcessExtension/dom-input-element-is-user-edited ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22774 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34116 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11931 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28641 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33027 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38438 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/128893 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32773 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35163 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/ref-i31-eq.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36122 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34371 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
```diff
@@ -44,6 +44,8 @@
 namespace WebKit {
 using namespace WebCore;
 
+uint8_t RemoteCaptureSampleManager::RemoteAudio::s_realtimeThreadCount { 0 };
```
shouldn't this be an atomic? 

Thie can be accessed on multiple threads at once without synchronisation.

jyavenard commented on Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
```diff
@@ -229,7 +234,31 @@ void RemoteCaptureSampleManager::RemoteAudio::startThread()
+    // during heavy loads. Realtime threads have a runtime cost, so ensure no more than two realtime
+    // capture threads can be created simultaneously.
+    // FIXME: Coalesce these threads to allow a single realtime thread to service every RemoteAudio instance.
+    bool shouldCreateRealtimeThread = s_realtimeThreadCount < s_maximumConcurrentRealtimeThreads;
```
this seems racy, doesn't it matter?

jernoble replied:
It probably doesn't, but I can update this to an atomic compare and increment.

youennf replied:
Can we add `ASSERT(isMainRunLoop())` for now?

youennf commented on Source/WebCore/platform/audio/RealtimeAudioThread.cpp:
```diff
@@ -0,0 +1,72 @@
+        uint8_t numberOfRealtimeThreads = 0;
+        for (auto* thread : Thread::allThreads()) {
+            if (thread && thread->isRealtime())
+                ++numberOfRealtimeThreads;
```
Maybe this counting could be moved to Threading.h as a static like done previously in RemoteAudioDestinationProxy.

jernoble replied:
I'd like to make a bigger change that would let `Thread` be something that could be sub-classed. Right now much of the work done to configure the thread is done in `Thread::create()` so that logic would all have to be reproduced. That way, the counting could be something done in a `RealtimeAudioThread` subclass.

youennf reviewed (approved):
LGTM.
We will likely need to work on coalescing threads since sites like Google Meet will at least use 3 threads at least in the future: two microphone captures and one (or more?) AudioContext.
Also, we should probably be more proactive at stopping the realtime thread whenever a track gets muted or stopped, I'll check this further.

jernoble committed:
WebContent Process capture thread can become de-prioritized; leading to underruns
https://bugs.webkit.org/show_bug.cgi?id=293295
rdar://problem/151699897

Reviewed by Youenn Fablet.

The Audio IO thread in the GPU process is running at a realtime scheduling priority,
reading samples from a shared ring buffer for Web Audio, and storing samples into a
shared ring buffer for microphone capture. While the matching thread storing samples
for Web Audio in the Web Content process (owned by RemoteAudioDestinationManager) is
also running at a realtime priority, the matching thread reading microphone samples
from the shared ring buffer is running at a normal thread priority. This allows the
kernel to deprioritize and preempt that thread, which can lead to dropped microphone
samples when used as an input to Web Audio.

For now, the limit to the number of concurrent realtime audio threads is 3. Given
that its relatively unlikely to have three capture sessions running simultaneously
this is likely good enough, and in the future, we can and should coalesce all of
of these threads into a single audio thread.

* Source/WTF/wtf/Threading.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
(WebKit::RemoteCaptureSampleManager::RemoteAudio::stopThread):
(WebKit::RemoteCaptureSampleManager::RemoteAudio::startThread):
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.h:

Canonical link: https://commits.webkit.org/295769@main
